### ISSUE TITLE
[REVIEW]: A Julia package for bilevel optimization problems

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#1278]
**Submitting author:** @matbesancon (<a href="http://orcid.org/0000-0002-6284-3033">Mathieu Besançon</a>)
**Repository:** <a href="https://github.com/matbesancon/BilevelOptimization.jl" target ="_blank">https://github.com/matbesancon/BilevelOptimization.jl</a>
**Version:** v0.2.1
**Editor:** @poulson
**Reviewer:** @HaoZeke, @vissarion
**Archive:** <a href="https://doi.org/10.5281/zenodo.3249188" target="_blank">10.5281/zenodo.3249188</a>

## Status

[![status](http://joss.theoj.org/papers/78ebab36f82fc977eb78b989aeb2f972/status.svg)](http://joss.theoj.org/papers/78ebab36f82fc977eb78b989aeb2f972)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/78ebab36f82fc977eb78b989aeb2f972"><img src="http://joss.theoj.org/papers/78ebab36f82fc977eb78b989aeb2f972/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/78ebab36f82fc977eb78b989aeb2f972/status.svg)](http://joss.theoj.org/papers/78ebab36f82fc977eb78b989aeb2f972)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/matbesancon/BilevelOptimization.jl" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@HaoZeke & @vissarion, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.theoj.org/about#reviewer_guidelines. Any questions/concerns please let @poulson know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/matbesancon/BilevelOptimization.jl">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** v0.2.1
- [x] **Authorship:** Has the submitting author (@matbesancon) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/matbesancon/BilevelOptimization.jl">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** v0.2.1
- [x] **Authorship:** Has the submitting author (@matbesancon) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @HaoZeke, it looks like you're currently assigned as the reviewer for this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01278/joss.01278/10.21105.joss.01278.pdf)

**Comment by HaoZeke** [comment]
Hi, @matbesancon I've initiated the review process and hope to finish in the next couple of days. I will be opening pull requests and issues on the repository tracker, but I might also ask for clarifications here (typically with matters pertaining to the paper itself). In addition, for the `paper.md`, I request that a copy to make changes on [hackmd.io](https://hackmd.io) is made available to me and @vissarion  since it is not appropriate to make changes to it via pull requests.

**Comment by HaoZeke** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Attempting to check references...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- http://doi.org/10.1007/s00446-006-0020-y is OK
- http://doi.org/10.1109/TPWRS.2019.2892607 is OK

MISSING DOIs

- https://doi.org/10.1007/978-1-4613-0307-7_12 may be missing for title: Bilevel optimization: theory, algorithms and applications
- https://doi.org/10.1137/15m1020575 may be missing for title: JuMP: A modeling language for mathematical optimization
- https://doi.org/10.1287/trsc.35.4.345.10433 may be missing for title: A bilevel model for toll optimization on a multicommodity transportation network
- https://doi.org/10.1016/j.ejor.2019.01.059 may be missing for title: Toll Caps in Privatized Road Networks

INVALID DOIs

- None
```

**Comment by HaoZeke** [comment]
@matbesancon, please update the `paper.md` file to fix the missing DOIs.

**Comment by matbesancon** [comment]
The hackmd can be found there:
https://hackmd.io/s/B1e2UUNIE#

**Comment by matbesancon** [comment]
@HaoZeke the first DOI seems invalid (the preprint does not correspond to the article suggested), I'm fixing the others in https://github.com/matbesancon/BilevelOptimization.jl/pull/9

**Comment by matbesancon** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Attempting to check references...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- http://doi.org/10.1137/15m1020575 is OK
- http://doi.org/10.1287/trsc.35.4.345.10433 is OK
- http://doi.org/10.1007/s00446-006-0020-y is OK
- http://doi.org/10.1016/j.ejor.2019.01.059 is OK
- http://doi.org/10.1109/TPWRS.2019.2892607 is OK

MISSING DOIs

- https://doi.org/10.1007/978-1-4613-0307-7_12 may be missing for title: Bilevel optimization: theory, algorithms and applications

INVALID DOIs

- None
```

**Comment by matbesancon** [comment]
missing DOI still invalid

**Comment by HaoZeke** [comment]
Yeah, we'll just have to let that one go. It does not seem to correspond to any DOI. Even the ISSN number does not seem valid. @matbesancon

**Comment by matbesancon** [comment]
yup I just saw it is considered as a book and not article on google scholar now, I'll update the bib

**Comment by matbesancon** [comment]
@whedon build pdf
@whedon check references

**Comment by matbesancon** [comment]
@whedon build pdf

**Comment by matbesancon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01278/joss.01278/10.21105.joss.01278.pdf)

**Comment by matbesancon** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Attempting to check references...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- http://doi.org/10.1137/15m1020575 is OK
- http://doi.org/10.1287/trsc.35.4.345.10433 is OK
- http://doi.org/10.1007/s00446-006-0020-y is OK
- http://doi.org/10.1016/j.ejor.2019.01.059 is OK
- http://doi.org/10.1109/TPWRS.2019.2892607 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by poulson** [comment]
:wave: @HaoZeke, just checking in as it's been about a week since any comments were provided. I know that reviewing requires a significant amount of your (volunteered!) time and appreciate your commitment on this.

**Comment by HaoZeke** [comment]
Sorry I was hospitalized in a road accident. I'll be discharged today but have some appointments with an orthodontist... I might take a few days before I get home and am able to finish. 


________________________________

**Comment by matbesancon** [comment]
sorry to read that, great recover! :/

**Comment by poulson** [comment]
@HaoZeke I am so sorry to hear about the accident but am glad to hear you are okay. Please excuse the distraction of my message; your recovery is obviously much more important.

**Comment by HaoZeke** [comment]
Thank you @matbesancon and @poulson for your kind wishes and for waiting so long. I have finally recovered enough to complete this review. I believe everything is in order.. Sorry the final signing off was delayed so much.

**Comment by matbesancon** [comment]
Absolutely no trouble, glad to hear you're doing good

**Comment by poulson** [comment]
:wave: @vissarion -- it's been a little while and the other reviewer has provided their feedback. Would you be able to provide your input some time in the near future?

**Comment by vissarion** [comment]
@poulson sorry for being inactive here, I can do it by the end of next week if this is not an issue.

**Comment by vissarion** [comment]
Hello, and again sorry for the late review. It is now complete. I raised and issue here: https://github.com/matbesancon/BilevelOptimization.jl/issues/14 regarding documentation.

**Comment by matbesancon** [comment]
thanks, on it today

**Comment by vissarion** [comment]
@matbesancon thanks for addressing the issue on documentation. I closed https://github.com/matbesancon/BilevelOptimization.jl/issues/14 and finished my review check-list.

**Comment by matbesancon** [comment]
@vissarion thanks for the discussion, I will adapt the paper to also include the signatures of `build_blp_model`

**Comment by matbesancon** [comment]
@whedon generate pdf from branch doc-joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch doc-joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01278/joss.01278/10.21105.joss.01278.pdf)

**Comment by matbesancon** [comment]
@whedon generate pdf from branch doc-joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch doc-joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01278/joss.01278/10.21105.joss.01278.pdf)

**Comment by matbesancon** [comment]
@whedon generate pdf from branch doc-joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch doc-joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01278/joss.01278/10.21105.joss.01278.pdf)

**Comment by matbesancon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01278/joss.01278/10.21105.joss.01278.pdf)

**Comment by labarba** [comment]
👋  @HaoZeke, @vissarion — it looks like this submission is ready for you to take a second look now, and possibly make a recommendation?

**Comment by HaoZeke** [comment]
@labarba, having taken a second look, I believe all previous concerns raised were addressed and I would recommend that this be accepted.

**Comment by poulson** [comment]
Thank you, @HaoZeke!

@vissarion: you still hasn't checked the 'Performance' box. It would be great to get a response from you on this and whether or not you feel this submission is ready for acceptance.

**Comment by vissarion** [comment]
@poulson I recommend acceptance. However, I am not clicking the box because, as far as I understand, there are no performance claims by the submitted software to confirm so the box should be left unchecked.

**Comment by kyleniemeyer** [comment]
Hi @poulson, checking on on this review—it looks like both reviewers have given their ok.

**Comment by poulson** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01278/joss.01278/10.21105.joss.01278.pdf)

**Comment by poulson** [comment]
@matbesancon Would you mind fixing the following typo/grammatical/citation/clarification issues?

"at the two level" -> "at the two levels"

"with possibly a disjoint feasible set" -> "with a possibly disjoint feasible set"

"solving method" -> "solver"

"for such case" -> "for such cases"

"the two principal being" -> "with the two primary approaches being"
[Citations needed for what the two approaches roughly are.]

Clarification: "into a regular one"?
[Citation of Special Ordered Sets of type 1]

"forcing at least one of them to 0" -> "forcing at least one of them to zero"

"m the JuMP model" -> "m: the JuMP model,", etc., adding an "and" in the second-to-last item.

"Where CM is the complementarity" -> "where CM is the..."

"Each arc has an initial cost, the lower-level" ->
  "Each arc has an initial cost, and the lower-level"

"This problem can entirely be modeled using the framework presented above, a" ->
"This problem can be entirely modeled using the framework presented above, using a composite datatype defined in the package for holding all required data,"

Add commas and an "and" in the final list, such as: "The initial matrix of arc costs `init_cost`,"

"Even though BilevelOptimization.jl is thought for" -> "Even though BilevelOptimization.jl is typically for"

After doing so, could you register a DOI on zenodo and post it here?

**Comment by danielskatz** [comment]
👋 @matbesancon - this review is waiting on you to make changes and then to archive the repo.

**Comment by matbesancon** [comment]
apologies to all for the delays! I'll edit the paper with @poulson's comments

**Comment by matbesancon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01278/joss.01278/10.21105.joss.01278.pdf)

**Comment by matbesancon** [comment]
I'm registering a new version to trigger the DOI creation from Zenodo

**Comment by matbesancon** [comment]
The version of the software is v0.2.1, the corresponding DOI is 10.5281/zenodo.3249188

**Comment by poulson** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List all of Whedon's capabilities
@whedon commands

# Assign a GitHub user as the sole reviewer of this submission
@whedon assign @username as reviewer

# Add a GitHub user to the reviewers of this submission
@whedon add @username as reviewer

# Remove a GitHub user from the reviewers of this submission
@whedon remove @username as reviewer

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

# Change editorial assignment
@whedon assign @username as editor

# Set the software archive DOI at the top of the issue e.g.
@whedon set 10.0000/zenodo.00000 as archive

# Set the software version at the top of the issue e.g.
@whedon set v1.0.1 as version

# Open the review issue
@whedon start review

EDITORIAL TASKS

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Remind an author or reviewer to return to a review after a
# certain period of time (supported units days and weeks)
@whedon remind @reviewer in 2 weeks

# Ask Whedon to accept the paper and deposit with Crossref
@whedon accept

# Ask Whedon to check the references for missing DOIs
@whedon check references

```

**Comment by poulson** [comment]
@whedon set v0.2.1 as version

**Comment by whedon** [comment]
OK. v0.2.1 is the version.

**Comment by poulson** [comment]
@whedon set 10.5281/zenodo.3249188 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3249188" target="_blank">10.5281/zenodo.3249188</a> is the archive.

**Comment by poulson** [comment]
Congratulations, @matbesancon, I have recommended the paper for publication! (@openjournals/joss-eics)

**Comment by kyleniemeyer** [comment]
Hi @matbesancon, a few little things to fix before we publish:

- In your Zenodo archive, the summary/description says "A changelog is being generated, please check back in a few minutes."
- I made a couple of copy-edits to the paper in https://github.com/matbesancon/BilevelOptimization.jl/pull/20, could you merge that?

**Comment by matbesancon** [comment]
hi @kyleniemeyer thanks, I've updated the Zenodo description field, I responded on the repo for the edit, I'm not sure about the square bracket thing

**Comment by kyleniemeyer** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01278/joss.01278/10.21105.joss.01278.pdf)

**Comment by kyleniemeyer** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.1137/15m1020575 is OK
- 10.1287/trsc.35.4.345.10433 is OK
- 10.1007/s00446-006-0020-y is OK
- 10.1016/j.ejor.2019.01.059 is OK
- 10.1109/TPWRS.2019.2892607 is OK
- 10.1007/978-3-319-33461-5_9 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/808

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/808, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by kyleniemeyer** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1145718410208251904) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/809
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.01278](https://doi.org/10.21105/joss.01278)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by kyleniemeyer** [comment]
Congrats @matbesancon on your submission's publication in JOSS! Thanks to @HaoZeke and @vissarion for reviewing, and @poulson for editing.

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](http://joss.theoj.org/papers/10.21105/joss.01278/status.svg)](https://doi.org/10.21105/joss.01278)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.01278">
  <img src="http://joss.theoj.org/papers/10.21105/joss.01278/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: http://joss.theoj.org/papers/10.21105/joss.01278/status.svg
   :target: https://doi.org/10.21105/joss.01278
```
This is how it will look in your documentation:

[![DOI](http://joss.theoj.org/papers/10.21105/joss.01278/status.svg)](https://doi.org/10.21105/joss.01278)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: http://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.salsalabs.org/donate-to-joss

**Comment by matbesancon** [comment]
Thanks all for the reviews!
